### PR TITLE
Add Rust Environment module

### PR DESCRIPTION
**Pull request by makubacki**
Adds helpers for checking if a firmare environment is ready to build Rust.

The primary purpose of the functionality in rust_environment is to be used in a build plugin wrapper. However, the public functions may be used in other contexts as well.

### CODE REVIEW

File: docs/user/features/rust_environment.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+public functions may be used in other contexts as well. The following functions are available:
+
+- `run()` - Checks the current environment for Rust build support.
+  - The checks can be customized with the `custom_tool_checks` and `custom_tool_filters` parameters.
+- `get_workspace_toolchain_version()` - Returns the rust toolchain version specified in the workspace toolchain file.
+- `verify_workspace_rust_toolchain_is_installed()` - RVerifies the rust toolchain used in the workspace is available.
+
+### Integration Examples
+
+This section provides examples of how to use the functions available in `edk2toolext/codeql.py`.
```
Comment (Javagedes) [inline docs/user/features/rust_environment.md:21]:
```suggestion
This section provides examples of how to use the functions available in `edk2toolext/rust_environment.py`.
```

File: docs/user/features/rust_environment.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+## Usage
+
+The primary purpose of the functionality in `rust_environment` is to be used in a build plugin wrapper. However, the
+public functions may be used in other contexts as well. The following functions are available:
+
+- `run()` - Checks the current environment for Rust build support.
+  - The checks can be customized with the `custom_tool_checks` and `custom_tool_filters` parameters.
+- `get_workspace_toolchain_version()` - Returns the rust toolchain version specified in the workspace toolchain file.
+- `verify_workspace_rust_toolchain_is_installed()` - RVerifies the rust toolchain used in the workspace is available.
```
Comment (Javagedes) [inline docs/user/features/rust_environment.md:17]:
```suggestion
- `verify_workspace_rust_toolchain_is_installed()` - Verifies the rust toolchain used in the workspace is available.
```

File: edk2toolext/rust_environment.py [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+        return RustToolChainInfo(error=False, toolchain=toolchain_version)
+    except FileNotFoundError:
+        # If a file is not found. Do not check any further.
+        return RustToolChainInfo(error=True, toolchain=None)
+
+
+def run(
+    custom_tool_checks: dict[str, RustToolInfo] = {},
+    custom_tool_filters: List[CustomToolFilter] = [],
+) -> None:
```
Comment (Javagedes) [inline edk2toolext/rust_environment.py:298]:
```suggestion
) -> int:
```

File: docs/user/features/rust_environment.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+#### `run()`
+
+Call to check the environment for Rust build support:
+
+```python
+    import edk2toolext.rust_environment as rust_env
+
+    def Run(self):
+        rust_env.run()
```
Comment (Javagedes) [inline docs/user/features/rust_environment.md:31]: should set the return value to a variable in this example

File: edk2toolext/rust_environment.py [hunk 5]
```diff
@@ -0,0 +1,442 @@
+# @file rust_environment.py
```
Comment (Javagedes) [inline edk2toolext/rust_environment.py:1]: I feel like this file would make more sense in `edk2toolext/environment/*`. Thoughts? I'm not 100% sold on that thought.

File: edk2toolext/rust_environment.py [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+#     `RUST_ENV_CHECK_TOOL_EXCLUSIONS` with a comma separated list of the tools
+#     to exclude. For example, "rustup, cargo tarpaulin" would not require that
+#     those tools be installed.
+#
+# Copyright (c) Microsoft Corporation.
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+##
+"""Rust Environment Helper Functions.
+
+Functions to help check that a build environment is ready to build Rust code.
```
Comment (Javagedes) [inline edk2toolext/rust_environment.py:23]: Everything in the `""" ... """` is what shows up in the documentation at the top of the page (in API Reference) if you think any of the additional information you currently have in the above `#` documentation would be useful to a normal user.

File: edk2toolext/rust_environment.py [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+def verify_workspace_rust_toolchain_is_installed() -> RustToolChainInfo:
+    """Verifies the rust toolchain used in the workspace is available.
+
+    Note: This function does not use the toml library to parse the toml
+    file since the file is very simple and its not desirable to add the
+    toml module as a dependency.
+
+    Returns:
+        RustToolChainInfo: A tuple that indicates if the toolchain is
+        available and any the toolchain version if found.
```
Comment (Javagedes) [inline edk2toolext/rust_environment.py:243]: indent
```suggestion
            available and any the toolchain version if found.
```

File: edk2toolext/rust_environment.py [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+        toolchain=toolchain_version,
+    )
+
+
+def get_workspace_toolchain_version() -> RustToolChainInfo:
+    """Returns the rust toolchain version specified in the workspace toolchain file.
+
+    Returns:
+        RustToolChainInfo: The rust toolchain information. If an error
+        occurs, the error field will be True with no toolchain info.
```
Comment (Javagedes) [inline edk2toolext/rust_environment.py:280]:
```suggestion
            occurs, the error field will be True with no toolchain info.
```

File: edk2toolext/rust_environment.py [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+        f"{target_triple} rust-src"
+    )
+
+    return False
+
+
+def verify_workspace_rust_toolchain_is_installed() -> RustToolChainInfo:
+    """Verifies the rust toolchain used in the workspace is available.
+
+    Note: This function does not use the toml library to parse the toml
```
Comment (Javagedes) [inline edk2toolext/rust_environment.py:237]: The below suggestion will add a "pretty" note, see an example [here](http://127.0.0.1:8000/api/edk2_invocable/#edk2toolext.edk2_invocable.Edk2Invocable.GetWorkspaceRoot)
```
!!! note
    This function ...
```

File: edk2toolext/rust_environment.py [hunk 10]
```diff
@@ -0,0 +1,442 @@
+# @file rust_environment.py
+#
```
Comment (Javagedes) [inline edk2toolext/rust_environment.py:2]: Please add an args section to `run`, `get_workspace_toolchain_version` and `verify_workspace_rust_toolchain_is_installed`

e.g.

```
Args:
    custom_tool_checks (Dict[str, RustToolInfo], optional): ...description...
    custom_tool_filters (List[CustomToolFilter], optional): ...description...
```

**Review by Javagedes — APPROVED** [review]
Thanks for writing tests.]

File: edk2toolext/rust_environment.py (hunk 6 above)
Comment (makubacki) [reply edk2toolext/rust_environment.py:23]: Thanks, I forgot about that and inverted most of the content.

File: edk2toolext/rust_environment.py (hunk 5 above)
Comment (makubacki) [reply edk2toolext/rust_environment.py]: I think that makes sense. I moved it.

**Pull request merged**